src/win32/posix_w32.c

```diff
@@ -981,3 +981,55 @@ int p_inet_pton(int af, const char *src, void *dst)
 	errno = EINVAL;
 	return -1;
 }
+
+ssize_t p_pread(int fd, void *data, size_t size, off64_t offset)
+{
+	HANDLE fh;
+	DWORD rsize = 0;
+	OVERLAPPED ov = {0};
+	LARGE_INTEGER pos = {0};
+
+	/* TODO: Loop over to read in smaller chunks to support large reads */
+	if (!git__is_uint32(size)) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	pos.QuadPart = offset;
+	ov.Offset = pos.LowPart;
+	ov.OffsetHigh = pos.HighPart;
+	fh = (HANDLE)_get_osfhandle(fd);
+
+	if (ReadFile(fh, data, (DWORD)size, &rsize, &ov)) {
+		return (ssize_t)rsize;
+	}
+
+	set_errno();
+	return -1;
+}
+
+ssize_t p_pwrite(int fd, const void *data, size_t size, off64_t offset)
+{
+	HANDLE fh;
+	DWORD wsize = 0;
+	OVERLAPPED ov = {0};
+	LARGE_INTEGER pos = {0};
+
+	/* TODO: Loop over to write in smaller chunks to support large reads */
+	if (!git__is_uint32(size)) {
+		errno = EINVAL;
+		return -1;
+	}
```

Comment:
Good call to let the caller drive looping.

However, I am not sure if is a good idea to mimic POSIX limits in Win32 with the following code:
```
if (!git__is_ssizet(size) || git__add_int64_overflow(&final_offset, offset, (int64_t)size)) {
	errno = EINVAL;
	return -1;	
}
```

Reply:
given that `p_pwrite`/`p_pread` is trying to provide a POSIX-compatible implementation of `pwrite`/`pread`, it should in theory behave with the same semantics and impose similar restrictions. otherwise, the caller needs to know what semantics to expect of it, which kind of defeats the purpose of having compatibility wrappers.

(or did I misinterpret your comment?)

Reply:
I was wrong in trying to let Win32 behave differently. I agree having similar behavior is better that surprises due to differences in native implementations.